### PR TITLE
Sync epoch

### PR DESCRIPTION
**Pull request by vvarg229**
Closes https://github.com/nspcc-dev/neofs-testcases/issues/614

Tests result: 
https://http.t5.fs.neo.org/86C4P6uJC7gb5n3KkwEGpXRfdczubXyRNW5N9KeJRW73/330-1693812139/index.html
https://github.com/vvarg229/neofs-node/actions/runs/6063915255/job/16466308125

### CODE REVIEW

File: pytest_tests/steps/cluster_test_base.py [hunk 1]
```diff
@@ -17,18 +19,24 @@ def fill_mandatory_dependencies(self, cluster: Cluster, client_shell: Shell):
...
+    def wait_for_epochs_align(self, current_epoch_number: Optional[int] = None):
+        epoch.wait_for_epochs_align(self.shell, self.cluster, current_epoch_number)
 
     def get_epoch(self):
         return epoch.get_epoch(self.shell, self.cluster)
 
     def ensure_fresh_epoch(self):
         return epoch.ensure_fresh_epoch(self.shell, self.cluster)
+
+    @allure.step("Tick Epochs and wait for epochs align")
```
Comment (roman-khimov) [inline pytest_tests/steps/cluster_test_base.py:39]: `for epoch alignment` or `for epochs to align`

File: robot/resources/lib/python_keywords/epoch.py [hunk 2]
```diff
@@ -23,26 +23,30 @@
 
 @allure.step("Ensure fresh epoch")
 def ensure_fresh_epoch(
-    shell: Shell, cluster: Cluster, alive_node: Optional[StorageNode] = None
+        shell: Shell, cluster: Cluster, alive_node: Optional[StorageNode] = None
```
Comment (roman-khimov) [inline robot/resources/lib/python_keywords/epoch.py:26]: Whitespace change.

File: robot/resources/lib/python_keywords/epoch.py [hunk 3]
```diff
@@ -23,26 +23,30 @@
...
-        epochs.append(get_epoch(shell, cluster, node))
+        current_epoch = get_epoch(shell, cluster, node)
+        assert (
+                epoch_number is None or current_epoch > epoch_number
+        ), f"Epoch {current_epoch} wasn't ticked yet. Expected epoch > {epoch_number}"
+        epochs.append(current_epoch)
     unique_epochs = list(set(epochs))
     assert (
-        len(unique_epochs) == 1
+            len(unique_epochs) == 1
```
Comment (roman-khimov) [inline robot/resources/lib/python_keywords/epoch.py:49]: Whitespace

File: robot/resources/lib/python_keywords/epoch.py [hunk 4]
```diff
@@ -86,7 +90,7 @@ def tick_epoch(shell: Shell, cluster: Cluster, alive_node: Optional[StorageNode]
             neofs_adm_exec_path=NEOFS_ADM_EXEC,
             config_file=NEOFS_ADM_CONFIG_PATH,
         )
-        
+
```
Comment (roman-khimov) [inline robot/resources/lib/python_keywords/epoch.py:93]: Whitespace

File: robot/resources/lib/python_keywords/node_management.py [hunk 5]
```diff
@@ -169,7 +169,8 @@ def exclude_node_from_network_map(
     storage_node_set_status(node_to_exclude, status="offline")
 
     time.sleep(parse_time(MORPH_BLOCK_TIME))
-    tick_epoch(shell, cluster)
+    current_epoch = get_epoch(shell, cluster)
+    tick_epoch_and_wait(shell, cluster, current_epoch)
```
Comment (roman-khimov) [inline robot/resources/lib/python_keywords/node_management.py:173]: You don't use `current_epoch` here, so why don't you use a function that handles `get_epoch()` internally?

File: pytest_tests/steps/storage_object.py [hunk 6]
```diff
@@ -46,7 +46,8 @@ def delete_objects(
                 endpoint=cluster.default_rpc_endpoint,
             )
 
-    tick_epoch(shell, cluster)
+    current_epoch = get_epoch(shell, cluster)
+    tick_epoch_and_wait(shell, cluster, current_epoch)
```
Comment (roman-khimov) [inline pytest_tests/steps/storage_object.py:50]: And here as well.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
You can have a bit less commits, tick_and_wait changes can be made throughout the whole codebase in a single one.

File: pytest_tests/steps/cluster_test_base.py [hunk 7]
```diff
@@ -17,18 +19,24 @@ def fill_mandatory_dependencies(self, cluster: Cluster, client_shell: Shell):
         yield
 
     @allure.title("Tick {epochs_to_tick} epochs")
-    def tick_epochs(self, epochs_to_tick: int):
+    def tick_epochs(self, current_epoch: int, epochs_to_tick: int):
```
Comment (evgeniiz321) [inline pytest_tests/steps/cluster_test_base.py:22]: This method looks to me like tick_epochs_and_wait. Do we really need the second method? We could use self.get_epoch() as a default here if current_epoch is not specified

File: robot/resources/lib/python_keywords/epoch.py [hunk 8]
```diff
@@ -113,3 +117,11 @@ def tick_epoch(shell: Shell, cluster: Cluster, alive_node: Optional[StorageNode]
         gas=1,
     )
     sleep(parse_time(MAINNET_BLOCK_TIME))
+
+
+@allure.step("Tick Epoch and wait for epochs align")
+def tick_epoch_and_wait(shell: Shell, cluster: Cluster, current_epoch: Optional[int],
```
Comment (evgeniiz321) [inline robot/resources/lib/python_keywords/epoch.py:123]: why we have the same logic with epochs ticking in two places? here and in the cluster_test_base.py ?

File: pytest_tests/testsuites/session_token/test_object_session_token.py [hunk 9]
```diff
@@ -202,7 +202,7 @@ def test_session_token_expiration_flags(
                 session=session_token,
             )
 
-        self.tick_epochs(2)
+        self.tick_epochs_and_wait(2)
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pytest_tests/testsuites/session_token/test_object_session_token.py:205]: Different tests can have different requirements, so having a single current epoch is safer. And it's feasible in general, the next epoch update is expected to happen long after the current one once it's updated. If we have premature epoch ticks from IR that's a problem and failing a test in this case is appropriate behaviour.

File: pytest_tests/steps/cluster_test_base.py (hunk 1 above)
Comment (vvarg229) [reply pytest_tests/steps/cluster_test_base.py]: Fixed

File: robot/resources/lib/python_keywords/epoch.py (hunk 3 above)
Comment (vvarg229) [reply robot/resources/lib/python_keywords/epoch.py]: Fixed

File: robot/resources/lib/python_keywords/node_management.py (hunk 5 above)
Comment (vvarg229) [reply robot/resources/lib/python_keywords/node_management.py]: > You don't use `current_epoch` here, so why don't you use a function that handles `get_epoch()` internally?

Fixed

File: pytest_tests/steps/storage_object.py (hunk 6 above)
Comment (vvarg229) [reply pytest_tests/steps/storage_object.py]: > And here as well.

Fixed

**Comment by vvarg229** [comment]
> You can have a bit less commits, tick_and_wait changes can be made throughout the whole codebase in a single one.

Fixed

File: pytest_tests/steps/cluster_test_base.py [hunk 10]
```diff
@@ -16,19 +18,18 @@ def fill_mandatory_dependencies(self, cluster: Cluster, client_shell: Shell):
...
 
     def ensure_fresh_epoch(self):
         return epoch.ensure_fresh_epoch(self.shell, self.cluster)
+
+    @allure.step("Tick epochs and wait for epoch alignment")
+    def tick_epochs_and_wait(self, epochs_to_tick: int):
+        current_epoch = self.get_epoch()
+        for _ in range(epochs_to_tick):
+            self.tick_epoch()
+            epoch.wait_for_epochs_align(self.shell, self.cluster, current_epoch)
```
Comment (evgeniiz321) [inline pytest_tests/steps/cluster_test_base.py:35]: is it ok that current_epoch won't be really current? shouldn't it be updated inside the cycle?

Comment (vvarg229) [reply pytest_tests/steps/cluster_test_base.py:35]: > is it ok that current_epoch won't be really current? shouldn't it be updated inside the cycle?

That's right - we are comparing the current epoch and the epoch that will be after the ticking.

File: pytest_tests/testsuites/session_token/test_object_session_token.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (vvarg229) [reply pytest_tests/testsuites/session_token/test_object_session_token.py:205]: @cthulhu-rider let's continue the discussion in another issue?
I don't really understand how the ticking of an epoch can affect a session token.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**